features/share/impl/src/main/kotlin/io/element/android/features/share/impl/SharePresenter.kt

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2024 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.element.android.features.share.impl
+
+import android.content.Intent
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.MutableState
+import androidx.compose.runtime.mutableStateOf
+import dagger.assisted.Assisted
+import dagger.assisted.AssistedFactory
+import dagger.assisted.AssistedInject
+import io.element.android.libraries.architecture.AsyncAction
+import io.element.android.libraries.architecture.Presenter
+import io.element.android.libraries.architecture.runCatchingUpdatingState
+import io.element.android.libraries.core.bool.orFalse
+import io.element.android.libraries.matrix.api.MatrixClient
+import io.element.android.libraries.matrix.api.core.RoomId
+import io.element.android.libraries.mediaupload.api.MediaPreProcessor
+import io.element.android.libraries.mediaupload.api.MediaSender
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+class SharePresenter @AssistedInject constructor(
+    @Assisted private val intent: Intent,
+    private val appCoroutineScope: CoroutineScope,
+    private val shareIntentHandler: ShareIntentHandler,
+    private val matrixClient: MatrixClient,
+    private val mediaPreProcessor: MediaPreProcessor,
+) : Presenter<ShareState> {
+    @AssistedFactory
+    interface Factory {
+        fun create(intent: Intent): SharePresenter
+    }
+
+    private val shareActionState: MutableState<AsyncAction<List<RoomId>>> = mutableStateOf(AsyncAction.Uninitialized)
```

Comment:
Since this is already a property, wouldn't it be better to use `by` so we can avoid using `.value`? `CoroutineScope.share` wouldn't need it as a parameter either in that case.

Reply:
Sorry, I do not get your point about `by`. Can you share some code snippet please?

Reply:
So something like:

```kotlin
private var shareActionState by mutableStateOf<AsyncAction<List<RoomId>>>(AsyncAction.Uninitialized)

fun onRoomSelected(roomIds: List<RoomId>) {
    appCoroutineScope.share(intent, roomIds)
}

@Composable
override fun present(): ShareState {
    fun handleEvents(event: ShareEvents) {
        when (event) {
            ShareEvents.ClearError -> shareActionState = AsyncAction.Uninitialized
        }
    }

    return ShareState(
        shareAction = shareActionState,
        eventSink = { handleEvents(it) }
    )
}

...
```

That wouldn't be possible if the variable was part of the `present` function, but we can do it if it's a property in the presenter class.

Reply:
Strangely it does not compile here:

<img width="1131" alt="image" src="https://github.com/element-hq/element-x-android/assets/3940906/16374d97-13af-4292-8fdd-0303f2a8f7f9">

Reply:
because the type is not correct with `by`:

<img width="460" alt="image" src="https://github.com/element-hq/element-x-android/assets/3940906/a9debfa6-9bc1-4796-bf04-b2c32b186d01">

Reply:
Ah, true, I completely forgot about `runCatchingUpdatingState`, sorry. No need to change it then.